src/adaXT/random_forest/RandomForest.py

```diff
@@ -0,0 +1,355 @@
+
+    def fit(self, X: np.ndarray, Y: np.ndarray):
+        """
+        Function used to fit a forrest using many DecisionTrees for the given data
+
+        Parameters
+        ----------
+        X : np.ndarray
+            feature values
+        Y : np.ndarray
+            response values
+        """
+        if not self.bootstrap and self.max_samples:
+            raise AttributeError(
+                "Bootstrap can not be False while max_samples is set")
+
+        X, Y = self.check_input(X, Y)
+        self.features = SharedNumpyArray(X)
+        self.outcomes = SharedNumpyArray(Y)
+        self.n_obs, self.n_features = self.features._shape
+        if self.forrest_type == "Classification":
+            self.classes = np.unique(self.outcomes.read())
+        if self.max_samples is None:
+            self.max_samples = self.n_obs
+        self.trees = [
+            DecisionTree(
+                tree_type=self.forrest_type,
+                criteria=self.criterion,
+                max_depth=self.max_depth,
+                impurity_tol=self.impurity_tol,
+                min_samples_split=self.min_samples_split,
+                min_samples_leaf=self.min_samples_leaf,
+                min_improvement=self.min_improvement,
+                max_features=self.max_features,
+                splitter=self.splitter) for _ in range(
+                self.n_estimators)]
+
+        # Fit trees
+        self.__build_trees()
+
+        # Register that the forrest was succesfully fitted
+        self.forest_fitted = True
+
+        return self
+
+    # Function that returns an ndarray of random ints used as the
+    # sample_indices
+    def __get_sample_indices(self):
+        if self.bootstrap:
+            return np.random.randint(
+                low=0, high=self.n_obs, size=self.max_samples)
+        else:
+            return None
+
+    def predict(self, X: np.ndarray):
+        """
+        Predicts a y-value for given X values using the trees of the forrest.
+
+        Parameters
+        ----------
+        X : np.ndarray
+            (N, M) numpy array with features to predict
+
+        Returns
+        -------
+        np.ndarray
+            (N) numpy array with the prediction
+        """
+        if not self.forest_fitted:
+            raise AttributeError(
+                "The forrest has not been fitted before trying to call predict")
+
+        self.predict_values = SharedNumpyArray(X)
+
+        # Predict using all the trees, each column is the predictions from one
+        # tree
+        tree_predictions = self.__predict_trees()
+        self.predict_values.unlink()
+
+        if self.forrest_type == "Regression":
+            # Return the mean answer from all trees for each row
+            return np.mean(tree_predictions, axis=1)
+
+        elif self.forrest_type == "Classification":
+            # Return the element most voted for by trees for each row
+            # QUESTION: how to handle equal majority vote?
```

Comment:
Good question. I think its fine to have some well-defined behavior (e.g., always smaller class) and then mention this in the function description...

Reply:
Done.